### PR TITLE
Create stream resets table

### PR DESCRIPTION
**Pull request by alovew**
Migration to add stream resets table

### CODE REVIEW

**Comment by alovew** [comment]
I tried testing running this migration locally with `./gradlew :airbyte-db:lib:runConfigsMigration`. That command succeeded and from the output it looked like the migration was successfully applied, but I didn't see any change to my local db. Is there something else I need to add to this? cc: @benmoriceau @lmossman

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateResets.java
Comment (benmoriceau) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateResets.java:39]: Do we have limitation on the stream name size? I think it should be a field without limitation.

**Review by benmoriceau — COMMENTED** [review]
I think that we should generate the jooq objects as well. Is it something we should commit as well?

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateResets.java
Comment (alovew) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateResets.java:39]: the spec has it as a varchar but doesn't have a limit listed. i could make it a text field instead - what do you think?

Comment (benmoriceau) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateResets.java:39]: I think we should avoid having a limit because none is defined right now. So going for text type sounds right.

**Comment by lmossman** [comment]
@alovew A few notes:
1. I took a look at how the `runConfigsMigration` gradle command [is implemented](https://github.com/airbytehq/airbyte/blob/3d9a972d28e0c0cd6d9307e2bfc8da679b8ddb32/airbyte-db/lib/build.gradle#L42-L46), and you can see that it is calling the `MigrationDevCenter` java class with args `configs` and `migrate`. If you look at that class, you can see that with those inputs [the main method is calling "new ConfigsDatabaseMigrationDevCenter().runLastMigration()"](https://github.com/airbytehq/airbyte/blob/3d9a972d28e0c0cd6d9307e2bfc8da679b8ddb32/airbyte-db/lib/src/main/java/io/airbyte/db/instance/development/MigrationDevCenter.java#L122), which is actually [temporarily spinning up a postgresSQL container](https://github.com/airbytehq/airbyte/blob/3d9a972d28e0c0cd6d9307e2bfc8da679b8ddb32/airbyte-db/lib/src/main/java/io/airbyte/db/instance/development/MigrationDevCenter.java#L75) to perform the migration in, which is automatically destroyed at the end of the `try` block since the container is being created in a [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) fashion. This means that you will not see any effect on your local db from running that gradle command. In order to see the migration performed on your local db, you will need to build airbyte with these changes, and then deploy it locally.
2. For the JOOQ generation, remember that this creates classes in the `airbyte-db/jooq/build` directory, which is not tracked in git. In order to have java classes representing the records in this table, I think we usually define the class as a yaml in the `airbyte-config/models/src/main/resources/types/` directory, and then we run `SUB_BUILD=PLATFORM ./gradlew airbyte-config:models:generateJsonSchema2Pojo` to generate the class. 
However, this is just the way we have done things in the past, which I think is sort of a relic of when we used to store all configs as json blobs in a single table. Maybe it makes more sense to just use the generate jooq classes for this new table instead, since those should stay consistent with the db schema (though, this couples it more tightly to jooq)? Or we could use a simple [java record](https://docs.oracle.com/en/java/javase/14/language/records.html) to represent this? @benmoriceau @cgardens @jdpgrailsdev do any of you have thoughts here for how we should generate classes that represent the records in this new db table?
3. In this PR, we may want to also implement the java repository layer on top of this table that we will use to interact with it (i.e the class that will perform the inserts, deletions, etc. for this table). With just the raw migration being added, there isn't really any way for us to interact with this table at the moment, and implementing the persistence layer on top of it could potentially inform us of changes we would want to make to the schema.

**Comment by benmoriceau** [comment]
> 3\. In this PR, we may want to also implement the java repository layer on top of this table that we will use to interact with it (i.e the class that will perform the inserts, deletions, etc. for this table). With just the raw migration being added, there isn't really any way for us to interact with this table at the moment, and implementing the persistence layer on top of it could potentially inform us of changes we would want to make to the schema.

I think it is ok to split the PR. I th

**Comment by benmoriceau** [comment]
Sorry I close the PR when I wanted to delete my comment. @lmossman I think it is ok to split the table creation and the persistence layer, it is 2 different tickets and it could be merge independently.

**Comment by lmossman** [comment]
> Sorry I close the PR when I wanted to delete my comment. @lmossman I think it is ok to split the table creation and the persistence layer, it is 2 different tickets and it could be merge independently.

Okay that is probably fine, I don't feel too strongly about that point

**Comment by benmoriceau** [comment]
@alovew It seems that I was wrong about the generated class if they are in a build folder they should not be commited

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamResets.java
Comment (cgardens) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamResets.java:45]: i think our convention is that table names are singular.

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamResets.java
Comment (cgardens) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamResets.java:39]: i'm realizing that stream name isn't actually enough to identify a stream. it was raised when we were talking about the postgres state. we really need namespace and stream name. we have talked about assigning an id to streams so that we don't need a tuple, but for now, we need namespace + stream name (and add it to the index: connection_id, namespace, stream_name)

Comment (lmossman) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamResets.java:39]: We probably should place a uniqueness constraint on these two columns, since I don't believe we ever want to store multiple records in this table that both have the same connection ID and stream name.

**Review by cgardens — CHANGES_REQUESTED** [review]
Looks good! Requesting changes because I realized that there is one piece of information we missed in the initial design.

* I like the incremental approach here. Having the migration only keeps it straight forward.
* @lmossman i think the code generation runs automatically as part of the build. @jdpgrailsdev  was looking at some of this recently though, so i'll defer to him on what sounds right here.

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamResets.java
Comment (cgardens) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamResets.java:39]: good call.

**Comment by alovew** [comment]
@cgardens @lmossman @benmoriceau this is ready for another review

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamReset.java
Comment (lmossman) [inline airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamReset.java:53]: For both the `unique` constraint and the index, I think we should order it as `connectionId, streamName, streamNamespace`, because as I mentioned above its possible for the namespace to be null

**Review by lmossman — DISMISSED** [review]
Just a couple small comments about namespace being nullable

**Comment by alovew** [comment]
@lmossman @benmoriceau @cgardens could I get another review on this?

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamReset.java
Comment (cgardens) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamReset.java]: huh. why does this matter? i had never taken this into consideration when building an index.

**Review by cgardens — APPROVED** [review]
lgtm! thanks for updating it!

File: airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamReset.java
Comment (lmossman) [reply airbyte-db/lib/src/main/java/io/airbyte/db/instance/configs/migrations/V0_39_1_001__CreateStreamReset.java]: I was thinking if there was a connection that did not use namespaces, then the queries on this table for that connection would be based only on `connectionId` and `streamName`. Because of this, I thought we would want this index to be ordered as `connectionId, streamName, streamNamespace`, so that queries that are only specifying `where` clauses on `connectionId` and `streamName` would still be able to take full advantage of the index

**Pull request merged**
